packages/babel-core/src/config/printer.js

```diff
@@ -0,0 +1,133 @@
+// @flow
+
+import type {
+  OptionsAndDescriptors,
+  UnloadedDescriptor,
+} from "./config-descriptors";
+
+// todo: Use flow enums when @babel/transform-flow-types supports it
+export const ChainFormatter = {
+  Programmatic: 0,
+  Config: 1,
+};
+
+type PrintableConfig = {
+  content: OptionsAndDescriptors,
+  type: $Values<typeof ChainFormatter>,
+  callerName: ?string,
+  filepath: ?string,
+  index: ?number,
+  envName: ?string,
+};
+
+const Formatter = {
+  title(
+    type: $Values<typeof ChainFormatter>,
+    callerName: ?string,
+    filepath: ?string,
+  ): string {
+    let title = "";
+    if (type === ChainFormatter.Programmatic) {
+      title = "programmatic options";
+      if (callerName) {
+        title += " from " + callerName;
+      }
+    } else {
+      // $FlowIgnore
+      title = "config " + filepath;
+    }
+    return title;
+  },
+  loc(index: ?number, envName: ?string): string {
+    let loc = "";
+    if (index != null) {
+      loc += `.overrides[${index}]`;
+    }
+    if (envName != null) {
+      loc += `.env["${envName}"]`;
+    }
+    return loc;
+  },
+
+  optionsAndDescriptors(opt: OptionsAndDescriptors) {
+    const content = { ...opt.options };
+    // overrides and env will be printed as separated config items
+    delete content.overrides;
+    delete content.env;
+    // resolve to descriptors
+    const pluginDescriptors = [...opt.plugins()];
+    if (pluginDescriptors.length) {
+      content.plugins = pluginDescriptors.map(d => descriptorToConfig(d));
+    }
+    const presetDescriptors = [...opt.presets()];
+    if (presetDescriptors.length) {
+      content.presets = [...presetDescriptors].map(d => descriptorToConfig(d));
+    }
+    return JSON.stringify(content, undefined, 2);
+  },
+};
+
+function descriptorToConfig(d: UnloadedDescriptor): string | {} | Array<mixed> {
+  let name = d.file?.request;
+  if (name == null) {
+    if (typeof d.value === "object") {
+      name = d.value;
+    } else if (typeof d.value === "function") {
+      // If the unloaded descriptor is a function, i.e. `plugins: [ require("my-plugin") ]`,
+      // we print the first 50 characters of the function source code and hopefully we can see
+      // `name: 'my-plugin'` in the source
```

Comment:
Good hack, although I don't believe this will be the case for most plugins?

Even testing on https://codesandbox.io/s/lively-star-xv8bt, it will just print the code instead, the default plugin on ASTExplorer wouldn't even show the name until char 85.

Reply:
Good catch. Well since an object has no key orders, I don't think there exists an integer N such that the plugin name is included in the `(0, N)` substring.

My initial thought here is to provide user at least something more descriptive/recognizable than a generic `[Function]` string. I have come up with several approach:

- Print Function name. It could not help much because we don't have such convention that the default export should be a named function
- Use regex to extract `name: "blablabla"` from function source. It doesn't work with dynamic name and has false positive.
- Actually load the descriptor. We have to initialize the API object earlier, and it may introduce breaking changes as we can end up loading a plugin twice.

So I end up with this approach and hope it could serve as an indicator so that users could know what this plugin is -- most of the time these are ad-hoc plugins not in `node_modules`.

Reply:
> Use regex to extract name: "blablabla" from function source. It doesn't work with dynamic name

I think it would be really rare/weird to have a dynamic name, it's just a string? Yeah I don't think it'll be an issue in general but for standalone/repl which is a diff usecase altogether.

Reply:
I think that this is something that we can revisit with experience: I believe that the hard-to-understand function name might not be a problem for a few reasons:
1. Most people use `"babel-plugin-foo"` or `require.resolve("babel-plugin-foo")`
2. `.slice(0, 50)` always includes the plugin function name (or at least the first 41 characters if it's longer). This can be useful for plugins directly defined in the config file, [as we do](https://github.com/babel/babel/blob/e498bee10f0123bb208baa228ce6417542a2c3c4/babel.config.js#L168).
3. If the plugin is an actual anonymous function and the `.slice(0, 50)` substring doesn't contain any useful information, the person can easily see in which config file it is defined and check what's used at that index in the `plugins` array directly in the config file.